### PR TITLE
251 random spatial crop transforms

### PR DESCRIPTION
**Pull request by Nic-Ma**
Partly fixes issue 251.

### Description
This PR implemented below transforms:
1. CastToType
2. CastToTyped
3. CenterSpatialCrop
4. CenterSpatialCropd
5. RandSizeSpatialCrop
6. RandSizeSpatialCropd
7. RandCenterSpatialCrop
8. RandCenterSpatialCropd
And fixed issues in:
1. RandGaussianNoise
2. RandGaussianNoised
3. RandAdjustContrastd

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: docs/source/transforms.rst [hunk 1]
```diff
@@ -135,9 +141,9 @@ Vanilla Transforms
     :members:
     :special-members: __call__
 
-`RandUniformPatch`
+`RandCenterSpatialCrop`
 ~~~~~~~~~~~~~~~~~~~~
```
Comment (wyli) [inline docs/source/transforms.rst:145]: this gives a warning 'WARNING: Title underline too short.'

File: docs/source/transforms.rst [hunk 2]
```diff
@@ -328,9 +352,9 @@ Dictionary-based Composables
     :members:
     :special-members: __call__
 
-`RandUniformPatchd`
+`RandCenterSpatialCropd`
 ~~~~~~~~~~~~~~~~~~~~~
```
Comment (wyli) [inline docs/source/transforms.rst:356]: this gives a warning 'WARNING: Title underline too short.'

File: monai/transforms/transforms.py [hunk 3]
```diff
@@ -877,8 +877,75 @@ def __call__(self, img):
...
+    def __init__(self, roi_size):
+        self.roi_size = roi_size
+
+    def __call__(self, img):
+        center = [i // 2 for i in img.shape[1:]]
+        cropper = SpatialCrop(roi_center=center, roi_size=self.roi_size, copy=False)
+        return cropper(img)
+
+
+class RandCenterSpatialCrop(Randomizable):
```
Comment (wyli) [inline monai/transforms/transforms.py:901]: i think the class name should be `RandLocationSpatialCrop` or `RandLocSpatialCrop`, to be consistent with `RandSizeSpatialCrop`

File: monai/transforms/transforms.py [hunk 4]
```diff
@@ -877,8 +877,75 @@ def __call__(self, img):
...
+        roi_size = get_valid_patch_size(img.shape, self.roi_size)
+        self.randomize(img.shape, roi_size)
+        return img[self._slices]
+
+
+class RandSizeSpatialCrop(Randomizable):
+    """
+    Crop image with random size ROI. It can crop at a random position as center
+    or at the image center. And allows to set the minimum size to limit the randomly
+    generated ROI. Suppose all the expected fields specified by `keys` have same shape.
```
Comment (wyli) [inline monai/transforms/transforms.py:927]: Suppose -> 'This transform assumes'

File: monai/transforms/transforms.py [hunk 5]
```diff
@@ -877,8 +877,75 @@ def __call__(self, img):
...
+        self.random_center = random_center
+
+    def randomize(self, img_size):
+        min_size = [self.min_size] * len(img_size) if not isinstance(self.min_size, (list, tuple)) else self.min_size
+        self.roi_size = [self.R.randint(low=min_size[i], high=img_size[i] + 1) for i in range(len(img_size))]
+
+    def __call__(self, img):
+        self.randomize(img.shape[1:])
+        if self.random_center:
+            cropper = RandCenterSpatialCrop(self.roi_size)
```
Comment (wyli) [inline monai/transforms/transforms.py:945]: this transform already covers the random location cases, perhaps we should merge this class and `randcenterspatialcrop` into a `randspatialcrop`? with boolean options of `random_centre` and `random_size`

File: monai/transforms/composables.py [hunk 6]
```diff
@@ -360,65 +383,40 @@ def __call__(self, data):
 
 
 class RandGaussianNoised(Randomizable, MapTransform):
-    """Add gaussian noise to image.
+    """Add gaussian noise to image. Suppose all the expected fields have same shape.
```
Comment (wyli) [inline monai/transforms/composables.py:386]: 'gaussian' -> 'Gaussian', 'Suppose' -> 'This transform assumes'

File: docs/source/highlights.md [hunk 7]
```diff
@@ -49,7 +49,7 @@ new_img = affine(image, spatial_size=(300, 400), mode='bilinear')
 ```
 
 ### 4. Randomly crop out batch images based on positive/negative ratio
-Medical image data volume may be too large to fit into GPU memory. A widely-used approach is to randomly draw small size data samples during training. MONAI currently provides uniform random sampling strategy as well as class-balanced fixed ratio sampling which may help stabilize the patch-based training process.
+Medical image data volume may be too large to fit into GPU memory. A widely-used approach is to randomly draw small size data samples during training. MONAI currently provides many general random sampling strategies as well as class-balanced fixed ratio sampling which may help stabilize the patch-based training process.
```
Comment (wyli) [inline docs/source/highlights.md:52]: remove 'many', and 'as well as' -> 'including'

File: monai/transforms/composables.py [hunk 8]
```diff
@@ -360,65 +383,40 @@ def __call__(self, data):
...
 
     Args:
         keys (hashable items): keys of the corresponding items to be transformed.
             See also: :py:class:`monai.transforms.compose.MapTransform`
+        prob (float): Probability to add gaussian noise.
```
Comment (wyli) [inline monai/transforms/composables.py:391]: 'gaussian' -> 'Gaussian'

File: monai/transforms/composables.py [hunk 9]
```diff
@@ -555,7 +553,7 @@ def __call__(self, data):
 
 class RandAdjustContrastd(Randomizable, MapTransform):
     """
-    dictionary-based wrapper of RandAdjustContrast.
+    dictionary-based version RandAdjustContrast.
```
Comment (wyli) [inline monai/transforms/composables.py:556]: if you change 'RandAdjustContrast' to ``` :py:class:`monai.transforms.transfroms.RandAdjustContrast` ``` it'll show a nice link in the doc webpage. (it's the same for the other transforms)

File: docs/source/transforms.rst (hunk 1 above)
Comment (Nic-Ma) [reply docs/source/transforms.rst:145]: fixed in the latest patch.
Thanks.

File: docs/source/transforms.rst (hunk 2 above)
Comment (Nic-Ma) [reply docs/source/transforms.rst:356]: fixed in the latest patch.
Thanks.

File: monai/transforms/transforms.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: changed in the latest patch.
Thanks.

File: monai/transforms/transforms.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py:927]: changed in the latest patch.
Thanks.

File: monai/transforms/transforms.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py:945]: I think users usually use RandLocSpatialCrop only and it's more clear to separate into smaller transforms for better API user experience and future extension.
Welcome a further discussion if necessary.
Thanks.

File: monai/transforms/composables.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/transforms/composables.py:386]: changed in the latest patch.
Thanks.

File: docs/source/highlights.md (hunk 7 above)
Comment (Nic-Ma) [reply docs/source/highlights.md:52]: changed in the latest patch.
Thanks.

File: monai/transforms/composables.py (hunk 8 above)
Comment (Nic-Ma) [reply monai/transforms/composables.py:391]: changed in the latest patch.
Thanks.

File: monai/transforms/composables.py (hunk 9 above)
Comment (Nic-Ma) [reply monai/transforms/composables.py:556]: changed for all the transforms in our codebase in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I updated the code according to your comments.
Could you please help review it again?
Thanks.

**Pull request merged**